components/keyrings/common/data/psi_memory_resource.cpp

```diff
@@ -0,0 +1,24 @@
+#include "psi_memory_resource.hpp"
+
+#include "sql/current_thd.h"
+#include "sql/sql_class.h"
+
+PSI_memory_key KEY_mem_keyring;
+
+void *psi_memory_resource::do_allocate(std::size_t bytes,
+                                       std::size_t /*alignment*/) {
+  if (current_thd) {
+    PSI_thread *owner = thd_get_psi(current_thd);
+    psi_memory_service->memory_alloc(KEY_mem_keyring, bytes, &owner);
```

Comment:
Still, do not understand this. Why do we do the allocation twice in case `current_thd != nullptr`? Once via memory service and  one more time via `operator new[]`.

Reply:
Please check the implementation of memory_alloc (pfs_memory_alloc_vc in storage/perfschema/pfs.cc). It doesn't allocate anything, it doesn't even have a parameter that would return a pointer address. We don't use its return value because it simply returns KEY_mem_keyring, the first parameter.

It only tracks allocations.